#  Section 1: Getting Started with Python Environments

To do data science in Python, you need a **coding environment**.  
There are a few options, and you should be aware of them before we start.


In [3]:
# Run this cell to check Python version
!python --version

Python 3.13.15


# Section 2: Refresh on Python Basics

**Objectives:**
- Recall variables, lists, loops, and conditionals.
- Practice basic Python commands inside Jupyter.


In [4]:
# Variables and types
a = 10
b = 3.14
c = "hello"
d = True

print(type(a), type(b), type(c), type(d))


<class 'int'> <class 'float'> <class 'str'> <class 'bool'>


In [ ]:
# Basic operations
x = 5
print("x squared:", x ** 2)
print("x divided by 2:", x / 2)
print("floor division:", x // 2)
print("modulo:", x % 2)


x squared: 25
x divided by 2: 2.5
floor division: 2
modulo: 1


In [ ]:
# Lists and indexing
numbers = [10, 20, 30, 40, 50]
print("First element:", numbers[0])
print("Last element:", numbers[-1])
print("Slice [1:3]:", numbers[1:3])

First element: 10
Last element: 50
Slice [1:3]: [20, 30]


In [ ]:
# Loops and conditionals
for num in numbers:
    if num > 25:
        print(num, "is greater than 25")
    else:
        print(num, "is not greater than 25")

10 is not greater than 25
20 is not greater than 25
30 is greater than 25
40 is greater than 25
50 is greater than 25


###  Exercises
1. Create a variable that stores your name. Print "Hello, [your name]!"  
2. Make a list of three numbers. Print the sum and average.  
3. Write a loop that prints whether each number is even or odd.  
4. Bonus: Write a function `square_list(lst)` that returns a list of squares.

In [ ]:
name = "Ryan"
if name == "Ryan":
    print("Hello, Ryan!")

x = 10
print("x sum:", x+2)
print("x average:", x/2)

numbers = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]

for num in numbers:
    if num % 2 == 0:
        print(num, "is even")
    else:
        print(num, "is odd")



Hello, Ryan!
x sum: 12
x average: 5.0
1 is odd
2 is even
3 is odd
4 is even
5 is odd
6 is even
7 is odd
8 is even
9 is odd
10 is even


# Section 3: Working with Pandas

**Objectives:**
- Load datasets into Pandas DataFrames.
- Save and reload data using CSVs.
- Understand DataFrame vs Series.

 Note: This dataset is **clean**. Real-world data is usually messy (missing values, typos, encodings). *Cleaning Data is 70 to 80 percent of the job*


In [5]:
import pandas as pd
from sklearn.datasets import load_iris

# Load dataset
iris_data = load_iris(as_frame=True)
iris_df = iris_data.frame

iris_df.head()


,sepal length (cm),sepal width (cm),petal length (cm),petal width (cm),target
0,5.1,3.5,1.4,0.2,0
1,4.9,3.0,1.4,0.2,0
2,4.7,3.2,1.3,0.2,0
3,4.6,3.1,1.5,0.2,0
4,5.0,3.6,1.4,0.2,0


In [6]:
# Save dataset to CSV
iris_df.to_csv("iris.csv", index=False)

# Load it back
iris = pd.read_csv("iris.csv")

iris.head()


,sepal length (cm),sepal width (cm),petal length (cm),petal width (cm),target
0,5.1,3.5,1.4,0.2,0
1,4.9,3.0,1.4,0.2,0
2,4.7,3.2,1.3,0.2,0
3,4.6,3.1,1.5,0.2,0
4,5.0,3.6,1.4,0.2,0


###  Exercises
1. Save the first 20 rows of `iris` to `iris_sample.csv`.  
2. Load `iris_sample.csv` into `iris_small`.  
3. Confirm `iris_small.shape` is `(20, 5)`.  
4. Check if `iris_small["petal width (cm)"]` is a Series or DataFrame.


In [7]:
subset = iris_df.head(20)
subset.to_csv("iris_sample.csv", index=False)
iris_small = pd.read_csv("iris_sample.csv")
print(iris_small.shape)
print(type(iris_small["petal width (cm)"]))

(20, 5)
<class 'pandas.core.series.Series'>


# Section 4: Exploring DataFrames

**Objectives:**
- Preview and inspect DataFrames.
- Learn what rows and columns represent.
- Explore dataset dimensions.


In [8]:
# Preview
iris.head()
iris.tail()
iris.head(10)

# Not Pandas but also very cool and useful
from google.colab import sheets
sheet = sheets.InteractiveSheet(df=iris)

ModuleNotFoundError: No module named 'google'

In [9]:
# Columns and data types
print("Column names:", iris.columns)
print("Data types:")
print(iris.dtypes)

Column names: Index(['sepal length (cm)', 'sepal width (cm)', 'petal length (cm)',
       'petal width (cm)', 'target'],
      dtype='object')
Data types:
sepal length (cm)    float64
sepal width (cm)     float64
petal length (cm)    float64
petal width (cm)     float64
target                 int64
dtype: object


In [10]:
# Shape of the dataset
print("Shape:", iris.shape)
print("Number of rows:", len(iris))
print("Number of columns:", len(iris.columns))

Shape: (150, 5)
Number of rows: 150
Number of columns: 5


In [11]:
# Inspect a single row
iris.iloc[0]


sepal length (cm)    5.1
sepal width (cm)     3.5
petal length (cm)    1.4
petal width (cm)     0.2
target               0.0
Name: 0, dtype: float64

###  Exercises
1. Use `.head(15)` to preview the first 15 rows.  
2. Print all column names.  
3. Confirm dataset has 150 rows with `.shape`.  
4. Select row 25 with `.iloc`. What does it represent?


In [12]:
iris.head(15)

,sepal length (cm),sepal width (cm),petal length (cm),petal width (cm),target
0,5.1,3.5,1.4,0.2,0
1,4.9,3.0,1.4,0.2,0
2,4.7,3.2,1.3,0.2,0
3,4.6,3.1,1.5,0.2,0
4,5.0,3.6,1.4,0.2,0
5,5.4,3.9,1.7,0.4,0
6,4.6,3.4,1.4,0.3,0
7,5.0,3.4,1.5,0.2,0
8,4.4,2.9,1.4,0.2,0
9,4.9,3.1,1.5,0.1,0


In [13]:
print("Column names:", iris.columns)
print(iris_df.shape)
iris.iloc[25]

Column names: Index(['sepal length (cm)', 'sepal width (cm)', 'petal length (cm)',
       'petal width (cm)', 'target'],
      dtype='object')
(150, 5)


sepal length (cm)    5.0
sepal width (cm)     3.0
petal length (cm)    1.6
petal width (cm)     0.2
target               0.0
Name: 25, dtype: float64

# Section 5: Selecting and Filtering Data

**Objectives:**
- Select rows with `.iloc`.
- Select columns by name.
- Use boolean filtering.


In [14]:
# Selecting rows by range
iris.iloc[0:5]
iris.iloc[10:20]

,sepal length (cm),sepal width (cm),petal length (cm),petal width (cm),target
10,5.4,3.7,1.5,0.2,0
11,4.8,3.4,1.6,0.2,0
12,4.8,3.0,1.4,0.1,0
13,4.3,3.0,1.1,0.1,0
14,5.8,4.0,1.2,0.2,0
15,5.7,4.4,1.5,0.4,0
16,5.4,3.9,1.3,0.4,0
17,5.1,3.5,1.4,0.3,0
18,5.7,3.8,1.7,0.3,0
19,5.1,3.8,1.5,0.3,0


In [15]:
# Selecting columns
iris["sepal length (cm)"]        # Single column (Series)
iris[["sepal length (cm)", "petal length (cm)"]]   # Multiple columns (DataFrame)

,sepal length (cm),petal length (cm)
0,5.1,1.4
1,4.9,1.4
2,4.7,1.3
3,4.6,1.5
4,5.0,1.4
...,...,...
145,6.7,5.2
146,6.3,5.0
147,6.5,5.2
148,6.2,5.4


In [16]:
# Boolean selection
iris[iris["petal length (cm)"] > 1.5]

iris[(iris["sepal length (cm)"] > 5.0) &
     (iris["petal length (cm)"] < 2.0)]


,sepal length (cm),sepal width (cm),petal length (cm),petal width (cm),target
0,5.1,3.5,1.4,0.2,0
5,5.4,3.9,1.7,0.4,0
10,5.4,3.7,1.5,0.2,0
14,5.8,4.0,1.2,0.2,0
15,5.7,4.4,1.5,0.4,0
16,5.4,3.9,1.3,0.4,0
17,5.1,3.5,1.4,0.3,0
18,5.7,3.8,1.7,0.3,0
19,5.1,3.8,1.5,0.3,0
20,5.4,3.4,1.7,0.2,0


### Exercises
1. Select the first 10 rows of `iris`.  
2. Select only `"sepal width (cm)"` and `"petal width (cm)"`.  
3. Create a DataFrame where `"sepal length (cm)" > 6.0`.  
4. Bonus: Select rows where `"petal length (cm)" > 1.5` AND `"sepal width (cm)" < 3.0`.


In [17]:
iris.iloc[0:10]

,sepal length (cm),sepal width (cm),petal length (cm),petal width (cm),target
0,5.1,3.5,1.4,0.2,0
1,4.9,3.0,1.4,0.2,0
2,4.7,3.2,1.3,0.2,0
3,4.6,3.1,1.5,0.2,0
4,5.0,3.6,1.4,0.2,0
5,5.4,3.9,1.7,0.4,0
6,4.6,3.4,1.4,0.3,0
7,5.0,3.4,1.5,0.2,0
8,4.4,2.9,1.4,0.2,0
9,4.9,3.1,1.5,0.1,0


In [ ]:
iris[["sepal width (cm)", "petal width (cm)"]]

,sepal width (cm),petal width (cm)
0,3.5,0.2
1,3.0,0.2
2,3.2,0.2
3,3.1,0.2
4,3.6,0.2
...,...,...
145,3.0,2.3
146,2.5,1.9
147,3.0,2.0
148,3.4,2.3


In [18]:
two_cols = iris_df[["sepal width (cm)", "petal width (cm)"]]
print(two_cols)

long_sepal = iris_df[iris_df["sepal length (cm)"] > 6.0]
print(long_sepal)

     sepal width (cm)  petal width (cm)
0                 3.5               0.2
1                 3.0               0.2
2                 3.2               0.2
3                 3.1               0.2
4                 3.6               0.2
..                ...               ...
145               3.0               2.3
146               2.5               1.9
147               3.0               2.0
148               3.4               2.3
149               3.0               1.8

[150 rows x 2 columns]
     sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)  \
50                 7.0               3.2                4.7               1.4   
51                 6.4               3.2                4.5               1.5   
52                 6.9               3.1                4.9               1.5   
54                 6.5               2.8                4.6               1.5   
56                 6.3               3.3                4.7               1.6   
..        

# Section 6: Reflection & Next Steps

**Objectives:**
- Reflect on today’s practice.
- Prepare for next lecture (Filtering & Transformation).

### Reflection Questions
1. What did you learn about DataFrames today?  
2. How did you use AI as a tool (if at all)?  
3. Where did you get stuck? How did you solve it?  

### Next Lecture Preview
- Filtering data with conditions.
- Transforming columns with `.apply()`.
- Grouping and aggregating data with `.groupby()`.


- 1. I learned Selecting/slicing DataFrame rows (.head(), indexing), Selecting single vs. multiple columns (single brackets → Series, double brackets → DataFrame), Boolean filtering with one condition and combining conditions with & Saving/loading CSVs and checking .shape

- 2. I used AI as a tool to help confirm the correct pandas syntax for each step — like how to save/load CSV files, check .shape, and write boolean filters (including combining conditions with &). Rather than just copying answers, I used it to get the framework/pattern for each type of operation, then implemented the code myself based on what the lab was asking.

- 3. I didn't get stuck on any single error so much as I wanted to build out the code structure myself once I understood the pattern from each question, rather than just copying a solution. So I used AI more as a framework/reference to confirm my approach (e.g., single vs. double brackets for Series vs. DataFrame, how to combine boolean conditions with &) and then implemented it on my own.
